### PR TITLE
Load shedding

### PR DESCRIPTION
**Pull request by Ladicek**
Related to #36543

### CODE REVIEW

**Comment by Ladicek** [comment]
Draft because this is very much work in progress. Sharing to maybe get some initial feedback.

If this is considered to be too niche for core Quarkus, I'd be fine with moving to Quarkiverse.

CC @ahus1 @vmuzikar @tmonney

**Review by cescoffier — COMMENTED** [review]
Looks great!

First, I like the overload algorithm you are using. Any reference on it? 
Then, we have a connection limiter in Quarkus already, should we deprecate this in favor of the load shedding?

Also, I think it needs to be tested with long-running connection (gRPC streams, SSE, web sockets). 

I'm also thinking that on overload, we may need to adjust the readiness Kube probe. WDYT?

**Comment by Ladicek** [comment]
The algorithm is a simplified version of https://github.com/Netflix/concurrency-limits/blob/master/concurrency-limits-core/src/main/java/com/netflix/concurrency/limits/limit/VegasLimit.java, I think there's a link in the javadoc already.

> Then, we have a connection limiter in Quarkus already, should we deprecate this in favor of the load shedding?

Ah do we? I had no idea. Where can I learn more?

> I think it needs to be tested with long-running connection (gRPC streams, SSE, web sockets).

Good point, I didn't try that at all. I'll check, but I doubt I'll see anything meaningful. The present implementation is heavily oriented to request/response style of interaction. Occasional stream likely won't do anything, and a streaming-heavy application would require a more involved implementation.

> I'm also thinking that on overload, we may need to adjust the readiness Kube probe. WDYT?

Hmm, I'm not sure about that. That would be super coarse-grained.

**Comment by Ladicek** [comment]
Just marked as ready for review. I fixed a couple of bugs in the implementation and added some rudimentary documentation.

The feature is not very heavily tested. @franz1981 do you think it would be possible to test this in a perf lab? I don't really know how that works, but I guess we have a test that stresses Quarkus application above its capacity, where this would help?

**Comment by Ladicek** [comment]
Rebased and fixed the conflict.

**Comment by Ladicek** [comment]
Rebased and fixed a few tiny issues. I believe this is ready now.

**Review by cescoffier — APPROVED** [review]
I think we should merge it and iterate.

File: extensions/load-shedding/runtime/pom.xml [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+    <parent>
+        <groupId>io.quarkus</groupId>
+        <artifactId>quarkus-load-shedding-parent</artifactId>
+        <version>999-SNAPSHOT</version>
+    </parent>
+
+    <artifactId>quarkus-load-shedding</artifactId>
+
+    <name>Quarkus - Load Shedding - Runtime</name>
```
Comment (gsmet) [inline extensions/load-shedding/runtime/pom.xml:15]: I think we need a description here.

File: extensions/load-shedding/runtime/src/main/resources/META-INF/quarkus-extension.yaml [hunk 2]
```diff
@@ -0,0 +1,13 @@
+---
+artifact: ${project.groupId}:${project.artifactId}:${project.version}
+name: "Load Shedding"
+metadata:
+  keywords:
+  - "fault-tolerance"
+  - "load-shedding"
+  guide: "https://quarkus.io/guides/load-shedding"
```
Comment (gsmet) [inline extensions/load-shedding/runtime/src/main/resources/META-INF/quarkus-extension.yaml:8]: isn't it `load-shedding-reference`?

**Review by gsmet — COMMENTED** [review]
I added two small comments. I don't mind you merging and addressing them later so that we avoid some more conflicts.

File: extensions/load-shedding/runtime/pom.xml (hunk 1 above)
Comment (Ladicek) [reply extensions/load-shedding/runtime/pom.xml:15]: Will add.

File: extensions/load-shedding/runtime/src/main/resources/META-INF/quarkus-extension.yaml (hunk 2 above)
Comment (Ladicek) [reply extensions/load-shedding/runtime/src/main/resources/META-INF/quarkus-extension.yaml:8]: Ouch! Good point, will change.

File: extensions/load-shedding/runtime/pom.xml (hunk 1 above)
Comment (Ladicek) [reply extensions/load-shedding/runtime/pom.xml:15]: Done.

File: extensions/load-shedding/runtime/src/main/resources/META-INF/quarkus-extension.yaml (hunk 2 above)
Comment (Ladicek) [reply extensions/load-shedding/runtime/src/main/resources/META-INF/quarkus-extension.yaml]: Done.

**Review by gsmet — APPROVED** [review]

**Pull request merged**